File: src/ipc/commands.ts
// Typed wrappers around Tauri commands (src-tauri/src/ipc/commands.rs).
import { invoke } from "@tauri-apps/api/core";
import type {
  AgentInfo,
  AgentSession,
  CommitDetail,
  DiffSides,
  GraphRow,
  Hunk,
  RebaseCommit,
  RepoMeta,
  RepoStatus,
  ForgeInfo,
  ForgeIssue,
  ForgePull,
  GitflowConfig,
  LfsStatus,
  Remote,
  SigningConfig,
  StashEntry,
  Submodule,
  TermSession,
  Worktree,
} from "../types/git";

export const listRepos = () => invoke<RepoMeta[]>("list_repos");

export const openRepo = (path: string) => invoke<RepoMeta>("open_repo", { path });

export const removeRepo = (id: number) => invoke<void>("remove_repo", { id });

export const setFavorite = (id: number, favorite: boolean) =>
  invoke<void>("set_favorite", { id, favorite });

export const repoStatus = (path: string) => invoke<RepoStatus>("repo_status", { path });

export const graphLoad = (path: string, limit?: number) =>
  invoke<GraphRow[]>("graph_load", { path, limit });

export const commitDetail = (path: string, sha: string) =>
  invoke<CommitDetail>("commit_detail", { path, sha });

export const commitDiff = (path: string, sha: string, file: string) =>
  invoke<string>("commit_diff", { path, sha, file });

export const wipDiff = (path: string, file: string) =>
  invoke<string>("wip_diff", { path, file });

export const diffSides = (path: string, sha: string, file: string) =>
  invoke<DiffSides>("diff_sides", { path, sha, file });

export const fileHunks = (path: string, file: string, staged: boolean) =>
  invoke<Hunk[]>("file_hunks", { path, file, staged });
export const applyHunk = (path: string, patch: string, reverse: boolean) =>
  invoke<void>("apply_hunk", { path, patch, reverse });

export const stage = (path: string, files: string[]) => invoke<void>("stage", { path, files });
export const unstage = (path: string, files: string[]) => invoke<void>("unstage", { path, files });
export const stageAll = (path: string) => invoke<void>("stage_all", { path });
export const unstageAll = (path: string) => invoke<void>("unstage_all", { path });
export const commit = (path: string, message: string, amend: boolean) =>
  invoke<void>("commit", { path, message, amend });

export const listBranches = (path: string) => invoke<string[]>("list_branches", { path });
export const checkout = (path: string, refname: string) =>
  invoke<void>("checkout", { path, refname });
export const createBranch = (path: string, name: string, checkout: boolean) =>
  invoke<void>("create_branch", { path, name, checkout });
export const fetch = (path: string) => invoke<void>("fetch", { path });
export const pull = (path: string) => invoke<void>("pull", { path });
export const push = (path: string) => invoke<void>("push", { path });

export const merge = (path: string, refname: string) =>
  invoke<void>("merge", { path, refname });
export const createBranchAt = (path: string, name: string, at: string, checkout: boolean) =>
  invoke<void>("create_branch_at", { path, name, at, checkout });
export const branchDelete = (path: string, name: string, force: boolean) =>
  invoke<void>("branch_delete", { path, name, force });
export const branchRename = (path: string, oldName: string, newName: string) =>
  invoke<void>("branch_rename", { path, old: oldName, new: newName });
export const tagCreate = (path: string, name: string, at: string) =>
  invoke<void>("tag_create", { path, name, at });
export const tagDelete = (path: string, name: string) =>
  invoke<void>("tag_delete", { path, name });

export const resolveConflict = (path: string, file: string, ours: boolean) =>
  invoke<void>("resolve_conflict", { path, file, ours });
export const launchDifftool = (path: string, file: string) =>
  invoke<void>("launch_difftool", { path, file });
export const launchMergetool = (path: string, file: string) =>
  invoke<void>("launch_mergetool", { path, file });
export const conflictContent = (path: string, file: string) =>
  invoke<string>("conflict_content", { path, file });
export const writeResolution = (path: string, file: string, content: string) =>
  invoke<void>("write_resolution", { path, file, content });
export const mergeContinue = (path: string) => invoke<void>("merge_continue", { path });
export const mergeAbort = (path: string) => invoke<void>("merge_abort", { path });
export const remoteList = (path: string) => invoke<Remote[]>("remote_list", { path });
export const remoteAdd = (path: string, name: string, url: string) =>
  invoke<void>("remote_add", { path, name, url });
export const remoteRemove = (path: string, name: string) =>
  invoke<void>("remote_remove", { path, name });
export const remoteRename = (path: string, oldName: string, newName: string) =>
  invoke<void>("remote_rename", { path, old: oldName, new: newName });
export const remoteSetUrl = (path: string, name: string, url: string) =>
  invoke<void>("remote_set_url", { path, name, url });

export const cherryPick = (path: string, sha: string) =>
  invoke<void>("cherry_pick", { path, sha });
export const revert = (path: string, sha: string) => invoke<void>("revert", { path, sha });
export const reset = (path: string, target: string, mode: string) =>
  invoke<void>("reset", { path, target, mode });
export const sequencerContinue = (path: string, kind: string) =>
  invoke<void>("sequencer_continue", { path, kind });
export const sequencerAbort = (path: string, kind: string) =>
  invoke<void>("sequencer_abort", { path, kind });
export const rebase = (path: string, onto: string) => invoke<void>("rebase", { path, onto });
export const rebaseContinue = (path: string) => invoke<void>("rebase_continue", { path });
export const rebaseAbort = (path: string) => invoke<void>("rebase_abort", { path });
export const rebaseTodoCommits = (path: string, base: string) =>
  invoke<RebaseCommit[]>("rebase_todo_commits", { path, base });
export const rebaseInteractive = (path: string, base: string, todo: string) =>
  invoke<void>("rebase_interactive", { path, base, todo });

export const getSetting = (key: string) => invoke<string | null>("get_setting", { key });
export const setSetting = (key: string, value: string) =>
  invoke<void>("set_setting", { key, value });

export const detectAgents = () => invoke<AgentInfo[]>("detect_agents");

export const newAgentSession = (path: string, agentId: string, branch: string) =>
  invoke<AgentSession>("new_agent_session", { path, agentId, branch });
export const agentSessions = () => invoke<AgentSession[]>("agent_sessions");
export const agentWrite = (id: string, data: string) => invoke<void>("agent_write", { id, data });
export const agentResize = (id: string, rows: number, cols: number) =>
  invoke<void>("agent_resize", { id, rows, cols });
export const agentKill = (id: string) => invoke<void>("agent_kill", { id });
export const agentBuffer = (id: string) => invoke<string>("agent_buffer", { id });

export const terminalOpen = (cwd: string, title: string) =>
  invoke<TermSession>("terminal_open", { cwd, title });
export const terminalList = () => invoke<TermSession[]>("terminal_list");
export const terminalWrite = (id: string, data: string) =>
  invoke<void>("terminal_write", { id, data });
export const terminalResize = (id: string, rows: number, cols: number) =>
  invoke<void>("terminal_resize", { id, rows, cols });
export const terminalKill = (id: string) => invoke<void>("terminal_kill", { id });
export const terminalBuffer = (id: string) => invoke<string>("terminal_buffer", { id });

export const submoduleList = (path: string) => invoke<Submodule[]>("submodule_list", { path });
export const submoduleUpdate = (path: string, sub: string | null, init: boolean) =>
  invoke<void>("submodule_update", { path, sub, init });
export const submoduleSync = (path: string) => invoke<void>("submodule_sync", { path });

export const lfsStatus = (path: string) => invoke<LfsStatus>("lfs_status", { path });
export const lfsPull = (path: string) => invoke<void>("lfs_pull", { path });
export const lfsTrack = (path: string, pattern: string) =>
  invoke<void>("lfs_track", { path, pattern });
export const lfsLock = (path: string, file: string) => invoke<void>("lfs_lock", { path, file });
export const lfsUnlock = (path: string, file: string) =>
  invoke<void>("lfs_unlock", { path, file });

export const signingConfig = (path: string) =>
  invoke<SigningConfig>("signing_config", { path });
export const setSigning = (path: string, sign: boolean, format: string, key: string) =>
  invoke<void>("set_signing", { path, sign, format, key });

export const gitflowStatus = (path: string) =>
  invoke<GitflowConfig>("gitflow_status", { path });
export const gitflowInit = (path: string) => invoke<void>("gitflow_init", { path });
export const gitflowStart = (path: string, kind: string, name: string) =>
  invoke<void>("gitflow_start", { path, kind, name });
export const gitflowFinish = (path: string, kind: string, name: string) =>
  invoke<void>("gitflow_finish", { path, kind, name });

export const forgeDetect = (path: string) => invoke<ForgeInfo>("forge_detect", { path });
export const forgeSetToken = (provider: string, token: string) =>
  invoke<void>("forge_set_token", { provider, token });
export const forgeClearToken = (provider: string) =>
  invoke<void>("forge_clear_token", { provider });
export const forgePulls = (path: string) => invoke<ForgePull[]>("forge_pulls", { path });
export const forgeIssues = (path: string) => invoke<ForgeIssue[]>("forge_issues", { path });
export const openExternal = (url: string) => invoke<void>("open_external", { url });

export const worktreeList = (path: string) => invoke<Worktree[]>("worktree_list", { path });
export const worktreeAdd = (path: string, name: string, create: boolean) =>
  invoke<string>("worktree_add", { path, name, create });
export const worktreeRemove = (path: string, wtPath: string, force: boolean) =>
  invoke<void>("worktree_remove", { path, wtPath, force });

export const stashList = (path: string) => invoke<StashEntry[]>("stash_list", { path });
export const stashSave = (path: string, message: string | null, untracked: boolean) =>
  invoke<void>("stash_save", { path, message, untracked });
export const stashApply = (path: string, id: string) =>
  invoke<void>("stash_apply", { path, id });
export const stashPop = (path: string, id: string) => invoke<void>("stash_pop", { path, id });
export const stashDrop = (path: string, id: string) => invoke<void>("stash_drop", { path, id });

export const watchRepo = (path: string) => invoke<void>("watch_repo", { path });

export const unwatchRepo = (path: string) => invoke<void>("unwatch_repo", { path });

File: src-tauri/src/ipc/commands.rs
//! Tauri command surface (SPEC §5.3). Thin wrappers; logic lives in db/git/watcher.
//!
//! Git-backed commands are `async` so Tauri runs them on the async runtime
//! instead of the main thread — a blocking git call (e.g. `fetch`/`pull`/`push`
//! doing network I/O) must never freeze the UI. The fast local-DB and watcher
//! commands stay synchronous.

use tauri::{AppHandle, State};

use crate::db::Db;
use crate::error::{AppError, AppResult};
use crate::git;
use crate::agents;
use crate::supervisor::{self, AgentSession, Supervisor};
use crate::terminal::{TermSession, Terminals};
use crate::forge::{self, Provider};
use crate::model::{
    AgentInfo, CommitDetail, DiffSides, ForgeInfo, ForgeIssue, ForgePull, GitflowConfig, GraphRow,
    Hunk, LfsStatus, RebaseCommit, RepoMeta, RepoStatus, SigningConfig, StashEntry, Submodule,
    Worktree,
};
use crate::watcher::{self, Watchers};

#[tauri::command]
pub fn list_repos(db: State<Db>) -> AppResult<Vec<RepoMeta>> {
    db.list()
}

#[tauri::command]
pub fn open_repo(path: String, db: State<Db>) -> AppResult<RepoMeta> {
    git::validate(&path)?;
    let name = std::path::Path::new(&path)
        .file_name()
        .and_then(|n| n.to_str())
        .unwrap_or("repo")
        .to_string();
    db.add_or_touch(&path, &name)
}

#[tauri::command]
pub fn remove_repo(id: i64, db: State<Db>) -> AppResult<()> {
    db.remove(id)
}

#[tauri::command]
pub fn set_favorite(id: i64, favorite: bool, db: State<Db>) -> AppResult<()> {
    db.set_favorite(id, favorite)
}

#[tauri::command]
pub async fn repo_status(path: String) -> AppResult<RepoStatus> {
    git::status(&path)
}

#[tauri::command]
pub async fn graph_load(path: String, limit: Option<usize>) -> AppResult<Vec<GraphRow>> {
    git::graph(&path, limit.unwrap_or(2000))
}

#[tauri::command]
pub async fn commit_detail(path: String, sha: String) -> AppResult<CommitDetail> {
    git::commit_detail(&path, &sha)
}

#[tauri::command]
pub async fn commit_diff(path: String, sha: String, file: String) -> AppResult<String> {
    git::commit_diff(&path, &sha, &file)
}

#[tauri::command]
pub async fn wip_diff(path: String, file: String) -> AppResult<String> {
    git::wip_diff(&path, &file)
}

#[tauri::command]
pub async fn diff_sides(path: String, sha: String, file: String) -> AppResult<DiffSides> {
    git::diff_sides(&path, &sha, &file)
}

#[tauri::command]
pub async fn file_hunks(path: String, file: String, staged: bool) -> AppResult<Vec<Hunk>> {
    git::file_hunks(&path, &file, staged)
}

#[tauri::command]
pub async fn apply_hunk(path: String, patch: String, reverse: bool) -> AppResult<()> {
    git::apply_hunk(&path, &patch, reverse)
}

#[tauri::command]
pub async fn stage(path: String, files: Vec<String>) -> AppResult<()> {
    git::stage(&path, &files)
}

#[tauri::command]
pub async fn unstage(path: String, files: Vec<String>) -> AppResult<()> {
    git::unstage(&path, &files)
}

#[tauri::command]
pub async fn stage_all(path: String) -> AppResult<()> {
    git::stage_all(&path)
}

#[tauri::command]
pub async fn unstage_all(path: String) -> AppResult<()> {
    git::unstage_all(&path)
}

#[tauri::command]
pub async fn commit(path: String, message: String, amend: bool) -> AppResult<()> {
    git::commit(&path, &message, amend)
}

#[tauri::command]
pub async fn list_branches(path: String) -> AppResult<Vec<String>> {
    git::list_branches(&path)
}

#[tauri::command]
pub async fn checkout(path: String, refname: String) -> AppResult<()> {
    git::checkout(&path, &refname)
}

#[tauri::command]
pub async fn create_branch(path: String, name: String, checkout: bool) -> AppResult<()> {
    git::create_branch(&path, &name, checkout)
}

#[tauri::command]
pub async fn fetch(path: String) -> AppResult<()> {
    git::fetch(&path)
}

#[tauri::command]
pub async fn pull(path: String) -> AppResult<()> {
    git::pull(&path)
}

#[tauri::command]
pub async fn push(path: String) -> AppResult<()> {
    git::push(&path)
}

#[tauri::command]
pub async fn merge(path: String, refname: String) -> AppResult<()> {
    git::merge(&path, &refname)
}

#[tauri::command]
pub async fn create_branch_at(
    path: String,
    name: String,
    at: String,
    checkout: bool,
) -> AppResult<()> {
    git::create_branch_at(&path, &name, &at, checkout)
}

#[tauri::command]
pub async fn branch_delete(path: String, name: String, force: bool) -> AppResult<()> {
    git::branch_delete(&path, &name, force)
}

#[tauri::command]
pub async fn branch_rename(path: String, old: String, new: String) -> AppResult<()> {
    git::branch_rename(&path, &old, &new)
}

#[tauri::command]
pub async fn tag_create(path: String, name: String, at: String) -> AppResult<()> {
    git::tag_create(&path, &name, &at)
}

#[tauri::command]
pub async fn tag_delete(path: String, name: String) -> AppResult<()> {
    git::tag_delete(&path, &name)
}

#[tauri::command]
pub async fn resolve_conflict(path: String, file: String, ours: bool) -> AppResult<()> {
    git::resolve_side(&path, &file, ours)
}

#[tauri::command]
pub async fn launch_difftool(path: String, file: String) -> AppResult<()> {
    git::launch_difftool(&path, &file)
}

#[tauri::command]
pub async fn launch_mergetool(path: String, file: String) -> AppResult<()> {
    git::launch_mergetool(&path, &file)
}

#[tauri::command]
pub async fn conflict_content(path: String, file: String) -> AppResult<String> {
    git::conflict_content(&path, &file)
}

#[tauri::command]
pub async fn write_resolution(path: String, file: String, content: String) -> AppResult<()> {
    git::write_resolution(&path, &file, &content)
}

#[tauri::command]
pub async fn merge_continue(path: String) -> AppResult<()> {
    git::merge_continue(&path)
}

#[tauri::command]
pub async fn merge_abort(path: String) -> AppResult<()> {
    git::merge_abort(&path)
}

#[tauri::command]
pub async fn cherry_pick(path: String, sha: String) -> AppResult<()> {
    git::cherry_pick(&path, &sha)
}

#[tauri::command]
pub async fn revert(path: String, sha: String) -> AppResult<()> {
    git::revert(&path, &sha)
}

#[tauri::command]
pub async fn reset(path: String, target: String, mode: String) -> AppResult<()> {
    git::reset(&path, &target, &mode)
}

#[tauri::command]
pub async fn sequencer_continue(path: String, kind: String) -> AppResult<()> {
    git::sequencer_continue(&path, &kind)
}

#[tauri::command]
pub async fn sequencer_abort(path: String, kind: String) -> AppResult<()> {
    git::sequencer_abort(&path, &kind)
}

#[tauri::command]
pub async fn rebase(path: String, onto: String) -> AppResult<()> {
    git::rebase(&path, &onto)
}

#[tauri::command]
pub async fn rebase_continue(path: String) -> AppResult<()> {
    git::rebase_continue(&path)
}

#[tauri::command]
pub async fn rebase_abort(path: String) -> AppResult<()> {
    git::rebase_abort(&path)
}

#[tauri::command]
pub async fn rebase_todo_commits(path: String, base: String) -> AppResult<Vec<RebaseCommit>> {
    git::rebase_todo_commits(&path, &base)
}

#[tauri::command]
pub async fn rebase_interactive(path: String, base: String, todo: String) -> AppResult<()> {
    git::rebase_interactive(&path, &base, &todo)
}

#[tauri::command]
pub fn detect_agents() -> Vec<AgentInfo> {
    agents::detect_agents()
}

#[tauri::command]
pub fn new_agent_session(
    app: AppHandle,
    supervisor: State<Supervisor>,
    db: State<Db>,
    path: String,
    agent_id: String,
    branch: String,
) -> AppResult<AgentSession> {
    let agent = agents::detect_agents()
        .into_iter()
        .find(|a| a.id == agent_id && a.available)
        .ok_or_else(|| AppError::Msg(format!("agent '{agent_id}' not available")))?;
    let command = agent.path.clone().unwrap_or_else(|| agent.command.clone());
    let worktree = git::worktree_add(&path, &branch, true)?;

    // Claude Code: inject status hooks via --settings so the UI gets live status.
    let mut args: Vec<String> = Vec::new();
    let mut status_file: Option<String> = None;
    if agent.id == "claude" {
        let stamp = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .map(|d| d.as_nanos())
            .unwrap_or(0);
        let dir = std::env::temp_dir();
        let sf = dir.join(format!("gitmage-agent-{stamp}.status"));
        let settings = dir.join(format!("gitmage-agent-{stamp}.settings.json"));
        let sf_str = sf.to_string_lossy().into_owned();
        std::fs::write(&settings, supervisor::claude_hooks_settings(&sf_str))
            .map_err(|e| AppError::Msg(format!("write settings: {e}")))?;
        args.push("--settings".into());
        args.push(settings.to_string_lossy().into_owned());
        status_file = Some(sf_str);
    }

    // Optional setup commands (Preferences > Agents): run them in the new worktree
    // before the agent by wrapping the spawn in a shell script that execs the agent.
    let setup = db.get_setting("agent.setup")?.unwrap_or_default();
    let setup_lines: Vec<&str> = setup
        .lines()
        .map(str::trim)
        .filter(|l| !l.is_empty() && !l.starts_with('#'))
        .collect();

    let (command, args) = if setup_lines.is_empty() {
        (command, args)
    } else {
        let stamp = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .map(|d| d.as_nanos())
            .unwrap_or(0);
        let script = std::env::temp_dir().join(format!("gitmage-setup-{stamp}.sh"));
        let body = format!("#!/bin/sh\nset -e\n{}\nexec \"$@\"\n", setup_lines.join("\n"));
        std::fs::write(&script, body).map_err(|e| AppError::Msg(format!("write setup: {e}")))?;
        let mut wrapped = vec![script.to_string_lossy().into_owned(), command];
        wrapped.extend(args);
        ("sh".to_string(), wrapped)
    };

    let session = supervisor.start(
        &app,
        &agent.id,
        &agent.name,
        &command,
        &args,
        &branch,
        &worktree,
        status_file.as_deref(),
    )?;

    // Persist so the session is still listed after an app restart.
    let now = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map(|d| d.as_millis() as i64)
        .unwrap_or(0);
    let _ = db.save_session(&session, now);
    Ok(session)
}

#[tauri::command]
pub fn get_setting(db: State<Db>, key: String) -> AppResult<Option<String>> {
    db.get_setting(&key)
}

#[tauri::command]
pub fn set_setting(db: State<Db>, key: String, value: String) -> AppResult<()> {
    db.set_setting(&key, &value)
}

#[tauri::command]
pub fn agent_write(supervisor: State<Supervisor>, id: String, data: String) -> AppResult<()> {
    supervisor.write(&id, &data)
}

#[tauri::command]
pub fn agent_resize(
    supervisor: State<Supervisor>,
    id: String,
    rows: u16,
    cols: u16,
) -> AppResult<()> {
    supervisor.resize(&id, rows, cols)
}

#[tauri::command]
pub fn agent_kill(supervisor: State<Supervisor>, db: State<Db>, id: String) -> AppResult<()> {
    supervisor.kill(&id)?;
    // Killing (or removing an already-exited entry) drops it from persistence.
    let _ = db.delete_session(&id);
    Ok(())
}

#[tauri::command]
pub fn agent_sessions(supervisor: State<Supervisor>, db: State<Db>) -> AppResult<Vec<AgentSession>> {
    let live = supervisor.list();
    let live_ids: std::collections::HashSet<String> = live.iter().map(|s| s.id.clone()).collect();
    let mut out = live;
    // Persisted sessions not currently running (e.g. after a restart) → exited.
    for s in db.list_sessions()? {
        if !live_ids.contains(&s.id) {
            out.push(s);
        }
    }
    Ok(out)
}

#[tauri::command]
pub fn agent_buffer(supervisor: State<Supervisor>, id: String) -> String {
    supervisor.buffer(&id)
}

#[tauri::command]
pub fn terminal_open(
    terminals: State<Terminals>,
    app: AppHandle,
    cwd: String,
    title: String,
) -> AppResult<TermSession> {
    terminals.open(&app, &cwd, &title)
}

#[tauri::command]
pub fn terminal_write(terminals: State<Terminals>, id: String, data: String) -> AppResult<()> {
    terminals.write(&id, &data)
}

#[tauri::command]
pub fn terminal_resize(
    terminals: State<Terminals>,
    id: String,
    rows: u16,
    cols: u16,
) -> AppResult<()> {
    terminals.resize(&id, rows, cols)
}

#[tauri::command]
pub fn terminal_kill(terminals: State<Terminals>, id: String) -> AppResult<()> {
    terminals.kill(&id)
}

#[tauri::command]
pub fn terminal_list(terminals: State<Terminals>) -> Vec<TermSession> {
    terminals.list()
}

#[tauri::command]
pub fn terminal_buffer(terminals: State<Terminals>, id: String) -> String {
    terminals.buffer(&id)
}

#[tauri::command]
pub async fn submodule_list(path: String) -> AppResult<Vec<Submodule>> {
    git::submodule_list(&path)
}

#[tauri::command]
pub async fn submodule_update(path: String, sub: Option<String>, init: bool) -> AppResult<()> {
    git::submodule_update(&path, sub.as_deref(), init)
}

#[tauri::command]
pub async fn submodule_sync(path: String) -> AppResult<()> {
    git::submodule_sync(&path)
}

#[tauri::command]
pub async fn lfs_status(path: String) -> AppResult<LfsStatus> {
    git::lfs_status(&path)
}

#[tauri::command]
pub async fn lfs_pull(path: String) -> AppResult<()> {
    git::lfs_pull(&path)
}

#[tauri::command]
pub async fn lfs_track(path: String, pattern: String) -> AppResult<()> {
    git::lfs_track(&path, &pattern)
}

#[tauri::command]
pub async fn lfs_lock(path: String, file: String) -> AppResult<()> {
    git::lfs_lock(&path, &file)
}

#[tauri::command]
pub async fn lfs_unlock(path: String, file: String) -> AppResult<()> {
    git::lfs_unlock(&path, &file)
}

#[tauri::command]
pub async fn signing_config(path: String) -> AppResult<SigningConfig> {
    git::signing_config(&path)
}

#[tauri::command]
pub async fn set_signing(
    path: String,
    sign: bool,
    format: String,
    key: String,
) -> AppResult<()> {
    git::set_signing(&path, sign, &format, &key)
}

#[tauri::command]
pub async fn gitflow_status(path: String) -> AppResult<GitflowConfig> {
    git::gitflow_status(&path)
}

#[tauri::command]
pub async fn gitflow_init(path: String) -> AppResult<()> {
    git::gitflow_init(&path)
}

#[tauri::command]
pub async fn gitflow_start(path: String, kind: String, name: String) -> AppResult<()> {
    git::gitflow_start(&path, &kind, &name)
}

#[tauri::command]
pub async fn gitflow_finish(path: String, kind: String, name: String) -> AppResult<()> {
    git::gitflow_finish(&path, &kind, &name)
}

#[tauri::command]
pub async fn forge_detect(path: String) -> AppResult<ForgeInfo> {
    Ok(forge::detect(&path))
}

#[tauri::command]
pub async fn forge_set_token(provider: String, token: String) -> AppResult<()> {
    let p = Provider::from_key(&provider).ok_or_else(|| AppError::Msg("unknown provider".into()))?;
    forge::set_token(p, &token)
}

#[tauri::command]
pub async fn forge_clear_token(provider: String) -> AppResult<()> {
    let p = Provider::from_key(&provider).ok_or_else(|| AppError::Msg("unknown provider".into()))?;
    forge::clear_token(p)
}

fn forge_token(rr: &forge::RepoRef) -> AppResult<String> {
    forge::get_token(rr.provider)
        .ok_or_else(|| AppError::Msg("no token for this provider — add one in the forge panel".into()))
}

#[tauri::command]
pub async fn forge_pulls(path: String) -> AppResult<Vec<ForgePull>> {
    let rr = forge::require_ref(&path)?;
    let token = forge_token(&rr)?;
    forge::fetch_pulls(&rr, &token).await
}

#[tauri::command]
pub async fn forge_issues(path: String) -> AppResult<Vec<ForgeIssue>> {
    let rr = forge::require_ref(&path)?;
    let token = forge_token(&rr)?;
    forge::fetch_issues(&rr, &token).await
}

/// Open a URL in the user's default browser (used for PR/issue links).
#[tauri::command]
pub async fn open_external(url: String) -> AppResult<()> {
    // Only follow web links; never shell-execute arbitrary schemes.
    if !(url.starts_with("https://") || url.starts_with("http://")) {
        return Err(AppError::Msg("refusing to open non-http URL".into()));
    }
    #[cfg(target_os = "macos")]
    let mut cmd = std::process::Command::new("open");
    #[cfg(target_os = "windows")]
    let mut cmd = {
        let mut c = std::process::Command::new("cmd");
        c.args(["/C", "start", ""]);
        c
    };
    #[cfg(all(unix, not(target_os = "macos")))]
    let mut cmd = std::process::Command::new("xdg-open");

    cmd.arg(&url);
    cmd.spawn().map_err(|e| AppError::Msg(format!("open: {e}")))?;
    Ok(())
}

#[tauri::command]
pub async fn worktree_list(path: String) -> AppResult<Vec<Worktree>> {
    git::worktree_list(&path)
}

#[tauri::command]
pub async fn worktree_add(path: String, name: String, create: bool) -> AppResult<String> {
    git::worktree_add(&path, &name, create)
}

#[tauri::command]
pub async fn worktree_remove(path: String, wt_path: String, force: bool) -> AppResult<()> {
    git::worktree_remove(&path, &wt_path, force)
}

#[tauri::command]
pub async fn worktree_lock(path: String, wt_path: String, lock: bool) -> AppResult<()> {
    git::worktree_lock(&path, &wt_path, lock)
}

#[tauri::command]
pub async fn worktree_prune(path: String) -> AppResult<()> {
    git::worktree_prune(&path)
}

#[tauri::command]
pub async fn stash_list(path: String) -> AppResult<Vec<StashEntry>> {
    git::stash_list(&path)
}

#[tauri::command]
pub async fn stash_save(path: String, message: Option<String>, untracked: bool) -> AppResult<()> {
    git::stash_save(&path, message.as_deref(), untracked)
}

#[tauri::command]
pub async fn stash_apply(path: String, id: String) -> AppResult<()> {
    git::stash_apply(&path, &id)
}

#[tauri::command]
pub async fn stash_pop(path: String, id: String) -> AppResult<()> {
    git::stash_pop(&path, &id)
}

#[tauri::command]
pub async fn stash_drop(path: String, id: String) -> AppResult<()> {
    git::stash_drop(&path, &id)
}

#[tauri::command]
pub fn watch_repo(path: String, app: AppHandle, watchers: State<Watchers>) -> AppResult<()> {
    watcher::watch(&app, &watchers, &path)
}

#[tauri::command]
pub fn unwatch_repo(path: String, watchers: State<Watchers>) -> AppResult<()> {
    watcher::unwatch(&watchers, &path);
    Ok(())
}
